File: Lua/radix_sort.lua
function get_max(arr)
    local max_val = arr[1] or 0 
    for i = 2, #arr do
        if arr[i] > max_val then
            max_val = arr[i]
        end
    end
    return max_val
end

function counting_sort_radix(arr, exp)
    local n = #arr
    local output = {}  
    local count = {}   

    for i = 0, 9 do count[i] = 0 end

    for i = 1, n do
        local index = math.floor(arr[i] / exp) % 10
        count[index] = (count[index] or 0) + 1
    end

    for i = 1, 9 do
        count[i] = count[i] + (count[i - 1] or 0)
    end

    for i = n, 1, -1 do
        local index = math.floor(arr[i] / exp) % 10
        local pos = count[index]
        output[pos] = arr[i]
        count[index] = count[index] - 1
    end

    for i = 1, n do
        arr[i] = output[i] or arr[i] 
    end
end

function radix_sort(arr)
    if #arr == 0 then return end  

    local max_val = get_max(arr)
    local exp = 1
    while max_val / exp >= 1 do
        counting_sort_radix(arr, exp)
        exp = exp * 10
    end
end

local arr = {170, 45, 75, 90, 802, 24, 2, 66}
radix_sort(arr)
print(table.concat(arr, " "))

File: Python/radix_sort.py
def counting_sort_for_radix(arr, exp):
    n = len(arr)
    output = [0] * n
    count = [0] * 10
    for num in arr:
        index = abs(num // exp) % 10
        count[index] += 1
    for i in range(1, 10):
        count[i] += count[i - 1]
    i = n - 1
    while i >= 0:
        index = abs(arr[i] // exp) % 10
        output[count[index] - 1] = arr[i]
        count[index] -= 1
        i -= 1
    for i in range(n):
        arr[i] = output[i]

def radix_sort(arr):
    if len(arr) == 0:
        return arr

    neg_arr = [num for num in arr if num < 0]
    pos_arr = [num for num in arr if num >= 0]

    if pos_arr:
        max_val = max(pos_arr)
        exp = 1
        while max_val // exp > 0:
            counting_sort_for_radix(pos_arr, exp)
            exp *= 10

    if neg_arr:
        min_val = min(neg_arr)
        exp = 1
        while abs(min_val) // exp > 0:
            counting_sort_for_radix(neg_arr, exp)
            exp *= 10
        neg_arr.reverse()

    return neg_arr + pos_arr

arr = [3, 7, 4, 8, 6, 2, 1, 5, 0, -5, -2]
print(radix_sort(arr))